### DISCUSSION TITLE
Path parameter with List

### DISCUSSION THREAD

**Discussion by mariannasb** [fastapi/fastapi#6154] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.get(
    V4L2_ROUTE + "controls/{ctrl_list}",
    response_model=List[ControlValue],
    tags=[Tags.controls],
)
def get_controls(
    ctrl_list: List[str],
    params: CommonParams = Depends(),
) -> JSONResponse:
    """Get v4l2 controls"""
    return Camera.get_controls(ctrl_list, params.dev)
```

### Description

When trying to declare a route with a Path parameter which is a List I get the following error: `AssertionError: Path params must be of one of the supported types`

But from what I can see it should be possible to have Lists in the Path: https://stackoverflow.com/questions/2602043/rest-api-best-practice-how-to-accept-list-of-parameter-values-as-input

Fx: `http://our.api.com/Product/101404,7267261`

### Operating System

Linux

### FastAPI Version

0.85.0

### Python Version

Python 3.10.5

**Comment by JarroVGIT** [comment, marked as answer]
A `Path` param is always a `int` or a `str`. However, you could simply do the conversion yourself. The below example shows how to deal with a list of values like 'int,int,int' etc. Using pydantic, I will also perform basic validation, yay!

```python
from fastapi import Depends, FastAPI, HTTPException, Path
from pydantic import BaseModel, ValidationError

app = FastAPI()


class ListOfIds(BaseModel):
    ids: list[int]


def list_path_param(list_of_ids: str = Path()):
    try:
        ids = ListOfIds(ids=list_of_ids.split(","))
    except ValidationError:
        raise HTTPException(400, "IDs must be valid integers.")
    return ids


@app.get("/controls/{list_of_ids}")
async def get_list_of_ids(list_obj: ListOfIds = Depends(list_path_param)):
    return {"count": len(list_obj.ids), "ids": list_obj.ids}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

This will take the `str` and will split it (resulting in a `list`), and construct a Pydantic model with its parts (thus performing validation). The above will run as is.

**Comment by mariannasb** [comment]
cool, thanks

It could be nice to also add such an example to the docs/tutorials :)

**Comment by tiangolo** [comment]
Thanks for the help @JarroVGIT ! 🙇

And thanks for coming back to close the issue @mariannasb . ☕

I would suggest not using lists in path parameters. A path is normally expected to point to a unique resource (a unique thing), and if the list is in the wrong order it would probably still point to the same thing.

It's also not common (I've never seen that before), so it might be a bit confusing for other developers using your API.

The alternative I would suggest is query parameters. You can make them required, require a minimum number of items, etc.

**Comment by mariannasb** [comment]
Yes, I ended up using query parameters in the end.

Partly because that was my work around before I got this solution.

But mostly because it ended up making more sense with the rest of the API since I also needed to do a POST with a list (which ended up as a `Dict[str, ControlValue]`)

**Comment by tiangolo** [comment]
Ah, get it, great!
